<a href="https://colab.research.google.com/github/stefaniamartinezr/Anal-tica-de-Negocios/blob/main/2_%C3%81rbol_de_Decisi%C3%B3n.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##**Caso de Estudio**
Una entidad financiera (FINTECH) quiere implementar un modelo de árbol de decision para mejrar la preaprobación de créditos de consumo de sus solicitantes de este tipo de créditos. Para este proceso vamos a utilizar las variables:

*    Edad: Indica el número de años que posee una persona, o el tiempo que usted lleva en el sistema financiero. n
*    Ingresos:Engloba todos los ingresos que recibe una persona además si posee salario mensual (USD).
*    Egresos:
*    Monto (EAD):

0. Se procede con la carga de las librerias de trabajo

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import confusion_matrix

1. Se procede con la carga de los datos de trabajo

In [ ]:
nxl='/content/1. SolicitantesCrédito(USD) (2).xlsx'
XDB=pd.read_excel(nxl,sheet_name=1)

XD=XDB.iloc[:,[1,10,11,25]]
yd=XDB.iloc[:,32]
display(XD)

2. Se procede con la implementacion del modelo de arbol

In [ ]:
mar=DecisionTreeClassifier(criterion='gini',max_depth=4)
mar.fit(XD,yd)    #Aqui el modelo busca relación entrada-salida

#¿Y que fue lo que hizo el modelo?
ydp=mar.predict(XD)    #Esto es lo que prónostica el modelo

#Se construye la matriz de consfusión
cm=confusion_matrix(yd,ydp)
display(cm)
VN=cm[0,0]; FP=cm[0,1]; FN=cm[1,0]; VP=cm[1,1]
#Metricas de Desempeño
Ex=(VP + VN)/len(XD)    #1. Exactitud: Comportamiento General
print("La exactitud es:" ,Ex)
Sen=VP/(VP+FN)     #2. Sensibilidad: Como se comporta pronosticando PreAprobados
print("La sensibilidad es: ",Sen)
Spe=VN/(VN+FP)  #3. Especificidad: Como se comporta frente a los negativos
print("La especificidad")
Pre=VP/(VP+FP)  #4. Precisión: Como se comporta pronosticando PreAprobados
print("La Precición es: ",Pre)
PreNeg=VN/(VN+FN) #5. Precisión Negativa: Como se comporta pronósticando negativos
print("La Predicción Negativa es:",PreNeg)

array([[2301,  658],
       [ 644, 2239]])

La exactitud es: 0.7771311194796303
La sensibilidad es:  0.7766215747485259
La especificidad
La Precición es:  0.772868484639282
La Predicción Negativa es: 0.7813242784380305


3. Despliegue el Árbol de Decisión

In [ ]:
from sklearn.tree import export_graphviz     #Exporta los datos a un gráfico
from pydotplus import graph_from_dot_data    #Es un graficador

vs=["Edad", "Ingresos","Egresos", "Monto"]   #Titulos del Árbol
dot_data=export_graphviz(mar,feature_names=vs) #Exportar de números a gráfico
graph=graph_from_dot_data(dot_data)
graph.write_png('Arbol.png')


True

La variable más importante del arbol fue monto, despues ingresos y cuando las varaiables se repiten en el arbol, el arbol se conoce como ÁRBOL DE REEMPLAZAMIENTO.

Nodos puros poseen un indice gini de cero, **diferencia muy bien de preaprobados y prenegados**

**Analisis de Resultados**

De la base de datos se puede observar un total de 2883 solicitantes de crédito que poseen la categoria de prenegados, mientras que para la categoria de preaprobados la matriz cuenta con un total de 2883 créditos preaprobados. Delos 2959 datos, el modelo pronostico correctamente un total de 2301 datos. De los 2883 preparobados el modelo pronostico correctamente un total de 2239. El modelo logra identificar en un 77% los creditos prenegados (2301/2959), mientras que el modelo logró identiifcar en un 77,66% los creditos preaprobados (2239/2883).

Con respecto a la metricas podemos observar que el modelo logro una exactitud, 77.71%, lo que indica el buen comportamiento general del modelo frente a la clasificación de créditos en las dos categorias de preaprobación. Se destacan la especificidad y la precisión las cuales lograron valores en promedio por encima del 77%, lo que suera el limite inferior definido para este tipo de modelos de clasificación el cual se ubica en el 75%.

De acuerdo con el arbol de decision se destaca un nodo puro (10/0) el cual posee la siguiente regla de decisión, si una persona cumple con esta
SI Monto<=322 and Ing>376 AND Egre<=685 and Ing>644 (923/5),tendra una probabilidad de aprobación del 100%. Es importante destacar que los nodos puros poseen un gini=0 (El modelo diferencia muy bien los buenapagas).

A pesar de la no existencia de más nodos puros, se destacan los nodos extremos o las reglas del negocio extremas. Se destaca una segunda regla que logra un porcentaje de aprobacion 99%, y la cual posee la siguiente estructura: Si Monto <=322 and Ing<=376 and Monto<=178 and Ing<=232 (923/5). Por su parte, el nodo derecho posee un porcentaje de negación del 98%, en donde la regla se define: Si Monto>322 and Ing>896 and Monto>961 and Ing>896 and Monto>961 and Ing>1178 (7/576)